File: web/scheduler.js
/**
 * MatchMaker - Team formation + tournament format planner
 *
 * Exported functions:
 *   - formBalancedTeams(attendees, playersMap, teamSize) → { teams, spectators }
 *   - recommendFormat(teams, numCourts, matchDuration, totalTime) → plan
 *
 * Team formation rule:
 *   - T = floor(N / teamSize)
 *   - Top T ranked attendees = captains (1 per team)
 *   - Remaining shuffled, distributed round-robin into teams
 *   - Ranking metric: win rate (ties broken by total games desc, then random)
 */

function playerWinRate(p) {
  if (!p) return 0;
  const total = (p.wins || 0) + (p.draws || 0) + (p.losses || 0);
  return total > 0 ? p.wins / total : 0;
}

function playerTotalGames(p) {
  if (!p) return 0;
  return (p.wins || 0) + (p.draws || 0) + (p.losses || 0);
}

/* ─── Team Formation ────────────────────────────────────────── */
function formBalancedTeams(attendeeIds, playersMap, teamSize) {
  const n = attendeeIds.length;
  const numTeams = Math.floor(n / teamSize);

  if (numTeams < 2) {
    return { error: `至少需要 ${2 * teamSize} 人 (当前 ${n} 人)。`, teams: [], spectators: attendeeIds };
  }

  // Rank by win rate desc, then games played desc (more reliable), then random
  const jitter = {};
  attendeeIds.forEach(id => { jitter[id] = Math.random(); });

  const ranked = [...attendeeIds].sort((a, b) => {
    const wrA = playerWinRate(playersMap[a]);
    const wrB = playerWinRate(playersMap[b]);
    if (wrA !== wrB) return wrB - wrA;
    const gA = playerTotalGames(playersMap[a]);
    const gB = playerTotalGames(playersMap[b]);
    if (gA !== gB) return gB - gA;
    return jitter[a] - jitter[b];
  });

  const captains = ranked.slice(0, numTeams);
  const rest = shuffleCopy(ranked.slice(numTeams, numTeams * teamSize));
  const spectators = ranked.slice(numTeams * teamSize);

  const teams = captains.map((cap, i) => ({
    id: `t_${i}`,
    name: `Team ${i + 1}`,
    players: [cap],
  }));

  rest.forEach((pid, i) => {
    teams[i % numTeams].players.push(pid);
  });

  return { teams, spectators };
}

function shuffleCopy(arr) {
  const a = arr.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

/* ─── Round-robin pairings (Berger/circle method) ───────────── */
function roundRobinPairings(n) {
  const teams = Array.from({ length: n }, (_, i) => i);
  if (teams.length % 2 === 1) teams.push(-1);  // bye
  const m = teams.length;
  const rounds = [];
  let arr = teams.slice();
  for (let r = 0; r < m - 1; r++) {
    const roundPairs = [];
    for (let i = 0; i < m / 2; i++) {
      const a = arr[i], b = arr[m - 1 - i];
      if (a !== -1 && b !== -1) roundPairs.push([a, b]);
    }
    rounds.push(roundPairs);
    // rotate (fix first, move last to second)
    arr = [arr[0], arr[m - 1], ...arr.slice(1, m - 1)];
  }
  return rounds;
}

/* ─── Round-robin planner ───────────────────────────────────── */
function planRoundRobin(teams, numCourts, matchDuration, totalTime) {
  const T = teams.length;
  const rounds = roundRobinPairings(T);
  const maxSlots = Math.floor(totalTime / matchDuration);

  const schedule = [];
  let slotIdx = 0;

  for (let r = 0; r < rounds.length; r++) {
    const rp = rounds[r];
    for (let bs = 0; bs < rp.length; bs += numCourts) {
      const batch = rp.slice(bs, bs + numCourts);
      if (slotIdx >= maxSlots) {
        return { format: 'round-robin', schedule, fits: false, reason: '时间不足', slotsUsed: slotIdx };
      }
      schedule.push({
        phase: 'round-robin',
        round: r + 1,
        slot: slotIdx + 1,
        matches: batch.map(([a, b], i) => ({
          court: i + 1, team_a: a, team_b: b, kind: 'ranked',
        })),
      });
      slotIdx++;
    }
  }

  return { format: 'round-robin', schedule, fits: true, slotsUsed: slotIdx };
}

/* ─── Groups + Knockout planner ─────────────────────────────── */
function planGroupsKnockout(teams, numCourts, matchDuration, totalTime) {
  const T = teams.length;
  if (T < 4) return { fits: false, reason: '队伍太少' };

  let best = null;

  for (const groupSize of [4, 3]) {
    const numGroups = Math.floor(T / groupSize);
    if (numGroups < 2) continue;
    const remainder = T - numGroups * groupSize;
    const groupSizes = Array.from({ length: numGroups }, (_, i) =>
      groupSize + (i < remainder ? 1 : 0)
    );

    const totalAdvancing = 2 * numGroups;
    // Round down to power of 2
    let kn = 1;
    while (kn * 2 <= totalAdvancing) kn *= 2;
    if (kn < 2) continue;

    const plan = buildGroupsKnockout(groupSizes, kn, numCourts, matchDuration, totalTime);
    if (plan.fits) {
      if (!best || plan.slotsUsed < best.slotsUsed) best = plan;
    }
  }

  return best || { fits: false, reason: '时间不足以完成小组赛+淘汰赛' };
}

function buildGroupsKnockout(groupSizes, knSize, numCourts, matchDuration, totalTime) {
  const maxSlots = Math.floor(totalTime / matchDuration);

  // Assign team indices to groups sequentially (top teams get distributed later; for now seq)
  const groupTeamIds = [];
  let idx = 0;
  for (const size of groupSizes) {
    groupTeamIds.push(Array.from({ length: size }, (_, i) => idx + i));
    idx += size;
  }

  const schedule = [];
  let slotIdx = 0;

  // Group stage: compute rounds for each group, then interleave by round
  const groupRounds = groupTeamIds.map(ids => {
    const pairs = roundRobinPairings(ids.length);
    // Map local indices back to actual team indices
    return pairs.map(roundPairs =>
      roundPairs.map(([a, b]) => [ids[a], ids[b]])
    );
  });
  const maxRounds = Math.max(...groupRounds.map(g => g.length));

  for (let r = 0; r < maxRounds; r++) {
    const roundMatches = [];
    groupRounds.forEach(gr => {
      if (r < gr.length) gr[r].forEach(p => roundMatches.push(p));
    });
    for (let bs = 0; bs < roundMatches.length; bs += numCourts) {
      const batch = roundMatches.slice(bs, bs + numCourts);
      if (slotIdx >= maxSlots) return { fits: false, reason: '小组赛超时' };
      schedule.push({
        phase: 'group',
        round: r + 1,
        slot: slotIdx + 1,
        matches: batch.map(([a, b], i) => ({
          court: i + 1, team_a: a, team_b: b, kind: 'ranked',
        })),
      });
      slotIdx++;
    }
  }

  // Knockout stage — use placeholders (resolved later from results)
  let advancing = [];
  groupSizes.forEach((_, i) => {
    advancing.push(`G${i + 1}-1`);
    advancing.push(`G${i + 1}-2`);
  });
  advancing = advancing.slice(0, knSize);

  let current = advancing.slice();
  let knRound = 0;
  while (current.length > 1) {
    knRound++;
    const roundMatches = [];
    const nextRound = [];
    for (let i = 0; i < current.length; i += 2) {
      roundMatches.push([current[i], current[i + 1]]);
      nextRound.push(`KR${knRound}-M${Math.floor(i / 2) + 1}-W`);
    }

    // Reserve 1 court for friendly matches if we have more than 1 court
    const tournamentCourts = numCourts > 1 ? numCourts - 1 : 1;

    for (let bs = 0; bs < roundMatches.length; bs += tournamentCourts) {
      const batch = roundMatches.slice(bs, bs + tournamentCourts);
      if (slotIdx >= maxSlots) return { fits: false, reason: '淘汰赛超时' };
      const matches = batch.map(([a, b], i) => ({
        court: i + 1, team_a: a, team_b: b, kind: 'ranked',
      }));
      if (numCourts > batch.length) {
        // Add friendly court for eliminated teams
        matches.push({
          court: numCourts,
          team_a: null,
          team_b: null,
          kind: 'friendly',
        });
      }
      schedule.push({
        phase: 'knockout',
        round: `KR${knRound}`,
        slot: slotIdx + 1,
        matches,
      });
      slotIdx++;
    }

    current = nextRound;
  }

  return {
    format: 'groups-knockout',
    group_sizes: groupSizes,
    knockout_size: knSize,
    schedule,
    fits: true,
    slotsUsed: slotIdx,
  };
}

/* ─── Main recommender ──────────────────────────────────────── */
function recommendFormat(teams, numCourts, matchDuration, totalTime) {
  const T = teams.length;
  if (T < 2) return { error: '至少需要 2 支队伍', fits: false };

  if (T >= 4) {
    const gk = planGroupsKnockout(teams, numCourts, matchDuration, totalTime);
    if (gk.fits) return gk;
  }

  return planRoundRobin(teams, numCourts, matchDuration, totalTime);
}

/* ─── Random-fair fallback ──────────────────────────────────── */
/**
 * Used when neither groups+knockout nor round-robin fit in the time budget.
 * Discards the fixed weekly teams and instead schedules per-match teams,
 * prioritising fairness within each match (强强 vs 强强 / 弱弱 vs 弱弱 /
 * 强弱 vs 强弱) and equal participation across players.
 *
 * Strategy per slot per court:
 *   1. From players not yet used this slot, sort by (gameCount asc,
 *      winRate desc) so that we both equalise play counts AND cluster
 *      similar-skill players together.
 *   2. Take the first `playersPerMatch` of that order — these become
 *      the four-ish players who play this match. Because of the WR
 *      tie-break, when many players are tied on play count they tend
 *      to be selected as a similar-skill cohort.
 *   3. Within the cohort, snake-draft by win rate into `teamsPerMatch`
 *      teams. Snake draft balances the two teams so the match is fair
 *      internally.
 *
 * Returns the same shape as planRoundRobin/planGroupsKnockout, plus a
 * `teams` array containing all the per-match teams generated. Match
 * `team_a` / `team_b` reference these teams by numeric index.
 */
function planRandomFairFallback({
  attendeeIds, playersMap, teamSize, teamsPerMatch,
  numCourts, matchDuration, totalTime,
}) {
  const playersPerMatch = teamSize * teamsPerMatch;
  const maxSlots = Math.floor(totalTime / matchDuration);
  const n = attendeeIds.length;

  if (maxSlots <= 0) {
    return { fits: false, reason: '时间不足', format: 'random-fair' };
  }
  if (n < playersPerMatch) {
    return {
      fits: false,
      reason: `至少需要 ${playersPerMatch} 人，当前 ${n} 人`,
      format: 'random-fair',
    };
  }

  const courtsPerSlot = Math.min(numCourts, Math.floor(n / playersPerMatch));
  if (courtsPerSlot <= 0) {
    return { fits: false, reason: '人数不足以填满一个场地', format: 'random-fair' };
  }

  const wr = (id) => playerWinRate(playersMap[id]);

  // Per-player game count for this event
  const gameCount = {};
  attendeeIds.forEach(id => { gameCount[id] = 0; });

  const teams = [];      // global team list (one entry per generated match-side)
  const schedule = [];

  let teamCounter = 0;

  for (let slotIdx = 0; slotIdx < maxSlots; slotIdx++) {
    const usedThisSlot = new Set();
    const slotMatches = [];

    for (let courtIdx = 0; courtIdx < courtsPerSlot; courtIdx++) {
      const available = attendeeIds.filter(id => !usedThisSlot.has(id));
      if (available.length < playersPerMatch) break;

      // Sort by play count asc, win rate desc, jitter to break ties
      available.sort((a, b) => {
        const gd = gameCount[a] - gameCount[b];
        if (gd !== 0) return gd;
        const wd = wr(b) - wr(a);
        if (wd !== 0) return wd;
        return Math.random() - 0.5;
      });

      const cohort = available.slice(0, playersPerMatch);

      // Within the cohort, sort strictly by win rate desc and snake-draft.
      cohort.sort((a, b) => wr(b) - wr(a));

      // Snake draft into teamsPerMatch teams
      const teamPlayers = Array.from({ length: teamsPerMatch }, () => []);
      cohort.forEach((pid, i) => {
        const round = Math.floor(i / teamsPerMatch);
        const pos = i % teamsPerMatch;
        const teamIdx = round % 2 === 0 ? pos : (teamsPerMatch - 1 - pos);
        teamPlayers[teamIdx].push(pid);
      });

      // Materialise teams as global team objects and remember their indices
      const teamIndices = teamPlayers.map(players => {
        const idx = teams.length;
        teams.push({
          id: `t_${teamCounter++}`,
          name: `Team ${idx + 1}`,
          players,
        });
        return idx;
      });

      // For teamsPerMatch === 2 we have a single ranked match between
      // teamIndices[0] and teamIndices[1]. For teamsPerMatch > 2 we still
      // emit a single "match" record with the two teams in a multi-team
      // contest — keeping the schema consistent with the cup planners,
      // which represent each court as a single match between team_a/team_b.
      // (For now teamsPerMatch is effectively 2 in the current UI.)
      slotMatches.push({
        court: courtIdx + 1,
        team_a: teamIndices[0],
        team_b: teamIndices[1] !== undefined ? teamIndices[1] : teamIndices[0],
        kind: 'ranked',
      });

      // Bump usage
      cohort.forEach(pid => {
        usedThisSlot.add(pid);
        gameCount[pid] += 1;
      });
    }

    if (slotMatches.length > 0) {
      schedule.push({
        phase: 'random-fair',
        round: slotIdx + 1,
        slot: slotIdx + 1,
        matches: slotMatches,
      });
    }
  }

  if (schedule.length === 0) {
    return { fits: false, reason: '人数不足以填满一个场地', format: 'random-fair' };
  }

  return {
    format: 'random-fair',
    schedule,
    teams,
    fits: true,
    slotsUsed: schedule.length,
  };
}

/**
 * Wraps recommendFormat: try the cup formats first, then fall back to
 * the random-fair scheduler if neither fit. Used by the UI's
 * generateTeams() flow so the user always gets *some* schedule.
 */
function recommendFormatOrFallback({
  teams, attendeeIds, playersMap, teamSize, teamsPerMatch,
  numCourts, matchDuration, totalTime,
}) {
  const cupPlan = recommendFormat(teams, numCourts, matchDuration, totalTime);
  if (cupPlan.fits) return { plan: cupPlan, fallback: false };

  const fallback = planRandomFairFallback({
    attendeeIds, playersMap, teamSize, teamsPerMatch,
    numCourts, matchDuration, totalTime,
  });
  if (fallback.fits) {
    return { plan: fallback, fallback: true, cupReason: cupPlan.reason };
  }
  // Even the fallback failed — surface the original cup plan's error
  return { plan: cupPlan, fallback: false };
}

File: algorithm/scheduler.py
"""
MatchMaker - Weekly Cup Tournament Scheduler (Python prototype)

Workflow:
    1. Maintain a player database with points (W=3, D=1, L=0).
    2. Each event: select attendees, auto-form balanced teams based on ranking,
       recommend a tournament format (prefer groups + knockout), generate
       schedule, record results, update points.
"""

import math
import random
from collections import defaultdict


# ────────────────────────────────────────────────────────────────────────────
# Player Database
# ────────────────────────────────────────────────────────────────────────────

def new_player(name: str) -> dict:
    return {
        "name": name,
        "points": 0,
        "wins": 0,
        "draws": 0,
        "losses": 0,
        "events": 0,
    }


def apply_match_result(players_db: dict, team_a: list[str], team_b: list[str], result: str):
    """result in {'A', 'B', 'D'}."""
    for name in team_a + team_b:
        if name not in players_db:
            players_db[name] = new_player(name)

    if result == "A":
        for p in team_a:
            players_db[p]["points"] += 3
            players_db[p]["wins"] += 1
        for p in team_b:
            players_db[p]["losses"] += 1
    elif result == "B":
        for p in team_b:
            players_db[p]["points"] += 3
            players_db[p]["wins"] += 1
        for p in team_a:
            players_db[p]["losses"] += 1
    elif result == "D":
        for p in team_a + team_b:
            players_db[p]["points"] += 1
            players_db[p]["draws"] += 1


# ────────────────────────────────────────────────────────────────────────────
# Balanced team formation
# ────────────────────────────────────────────────────────────────────────────

def win_rate(p: dict) -> float:
    total = p.get("wins", 0) + p.get("draws", 0) + p.get("losses", 0)
    return p.get("wins", 0) / total if total > 0 else 0.0


def total_games(p: dict) -> int:
    return p.get("wins", 0) + p.get("draws", 0) + p.get("losses", 0)


def form_balanced_teams(
    attendees: list[str],
    players_db: dict,
    team_size: int,
    seed: int | None = None,
) -> dict:
    """
    Rule (from user):
      - T = floor(N / team_size) teams
      - Top T ranked attendees (by win rate) become captains, one per team
      - Remaining attendees are shuffled and distributed round-robin into teams
      - Ranking: win rate desc, games played desc, then random

    Returns: {"teams": [{"name", "players": [...]}], "spectators": [...]}
    """
    if seed is not None:
        random.seed(seed)

    n = len(attendees)
    num_teams = n // team_size
    if num_teams < 2:
        return {
            "error": f"Need at least {2 * team_size} players (got {n}).",
            "teams": [],
            "spectators": attendees,
        }

    jitter = {name: random.random() for name in attendees}

    def rank_key(name):
        p = players_db.get(name, new_player(name))
        return (-win_rate(p), -total_games(p), jitter[name])

    ranked = sorted(attendees, key=rank_key)
    captains = ranked[:num_teams]
    rest = ranked[num_teams : num_teams * team_size]
    spectators = ranked[num_teams * team_size :]

    random.shuffle(rest)

    teams = [{"name": f"Team {i+1}", "players": [captains[i]]} for i in range(num_teams)]

    # Distribute rest round-robin
    for i, player in enumerate(rest):
        teams[i % num_teams]["players"].append(player)

    return {"teams": teams, "spectators": spectators}


# ────────────────────────────────────────────────────────────────────────────
# Tournament format planning
# ────────────────────────────────────────────────────────────────────────────

def _round_robin_pairings(team_ids: list[int]) -> list[list[tuple[int, int]]]:
    """Berger/circle method. Returns list of rounds, each a list of (a, b) pairs.
    If odd number of teams, one sits out each round (marked as -1)."""
    n = len(team_ids)
    teams = list(team_ids)
    if n % 2 == 1:
        teams.append(-1)
    m = len(teams)
    rounds = []
    for r in range(m - 1):
        round_pairs = []
        for i in range(m // 2):
            a, b = teams[i], teams[m - 1 - i]
            if a != -1 and b != -1:
                round_pairs.append((a, b))
        rounds.append(round_pairs)
        teams = [teams[0]] + [teams[-1]] + teams[1:-1]
    return rounds


def plan_round_robin(teams: list[dict], num_courts: int, match_duration: int, total_time: int) -> dict:
    """Plan a single round-robin. Returns schedule + feasibility."""
    team_ids = list(range(len(teams)))
    rounds = _round_robin_pairings(team_ids)

    # Each "round" has matches that can be played in parallel on courts
    schedule = []  # list of time slots
    slot_idx = 0
    max_slots = total_time // match_duration

    for round_idx, round_pairs in enumerate(rounds):
        # Split round across slots if more matches than courts
        for batch_start in range(0, len(round_pairs), num_courts):
            batch = round_pairs[batch_start : batch_start + num_courts]
            if slot_idx >= max_slots:
                return {
                    "format": "round-robin",
                    "schedule": schedule,
                    "fits": False,
                    "reason": "超出时间限制",
                }
            schedule.append({
                "phase": "round-robin",
                "round": round_idx + 1,
                "slot": slot_idx + 1,
                "matches": [
                    {"court": i + 1, "team_a": a, "team_b": b, "kind": "ranked"}
                    for i, (a, b) in enumerate(batch)
                ],
            })
            slot_idx += 1

    return {
        "format": "round-robin",
        "schedule": schedule,
        "fits": True,
        "slots_used": slot_idx,
    }


def plan_groups_knockout(
    teams: list[dict], num_courts: int, match_duration: int, total_time: int
) -> dict:
    """Plan group stage + knockout. Tries several group sizes."""
    T = len(teams)
    if T < 4:
        return {"fits": False, "reason": "队伍太少，不适合小组赛"}

    # Decide group structure: prefer groups of 4, then 3
    # Number of groups = G, sizes balanced
    best = None
    for group_size in [4, 3]:
        num_groups = T // group_size
        if num_groups < 2:
            continue
        # Assign teams to groups (handle remainders by +1 to early groups)
        remainder = T - num_groups * group_size
        group_sizes = [group_size + (1 if i < remainder else 0) for i in range(num_groups)]

        # Advancing teams: top 2 per group (or top 1 if group_size == 3 and total advancing >= 4)
        advance_per_group = 2
        total_advancing = advance_per_group * num_groups

        # Must be power of 2 for clean knockout; drop extras from lowest seeds if not
        kn_size = 1
        while kn_size * 2 <= total_advancing:
            kn_size *= 2
        if kn_size < 2:
            continue

        plan = _build_groups_knockout(
            teams, group_sizes, kn_size, num_courts, match_duration, total_time
        )
        if plan["fits"]:
            if best is None or plan["slots_used"] < best["slots_used"]:
                best = plan

    if best:
        return best
    return {"fits": False, "reason": "时间不足以完成小组赛+淘汰赛"}


def _build_groups_knockout(teams, group_sizes, kn_size, num_courts, match_duration, total_time):
    T = len(teams)
    max_slots = total_time // match_duration

    # Assign team ids to groups sequentially (could randomize / snake later)
    group_team_ids = []
    idx = 0
    for size in group_sizes:
        group_team_ids.append(list(range(idx, idx + size)))
        idx += size

    schedule = []
    slot_idx = 0

    # --- Group stage ---
    # For each group, compute rounds; run groups in parallel on courts
    # Simpler: union all group-round pairings into time slots
    group_rounds = [_round_robin_pairings(g) for g in group_team_ids]
    max_rounds = max(len(gr) for gr in group_rounds)

    for r in range(max_rounds):
        # Collect all matches from round r of every group
        round_matches = []
        for g, gr in enumerate(group_rounds):
            if r < len(gr):
                for a, b in gr[r]:
                    round_matches.append((a, b))
        # Split across slots by court capacity
        for bs in range(0, len(round_matches), num_courts):
            batch = round_matches[bs : bs + num_courts]
            if slot_idx >= max_slots:
                return {"fits": False, "reason": "小组赛超时"}
            schedule.append({
                "phase": "group",
                "round": r + 1,
                "slot": slot_idx + 1,
                "matches": [
                    {"court": i + 1, "team_a": a, "team_b": b, "kind": "ranked"}
                    for i, (a, b) in enumerate(batch)
                ],
            })
            slot_idx += 1

    # --- Knockout stage ---
    # We don't know exact participants yet (depends on group results).
    # Represent matches as placeholders: ("G1-1", "G2-2") etc.
    advancing = []
    for g_idx, size in enumerate(group_sizes):
        advancing.append(f"G{g_idx+1}-1")
        advancing.append(f"G{g_idx+1}-2")
    # Trim to kn_size (lowest seeds removed)
    advancing = advancing[:kn_size]

    # Knockout bracket
    current = advancing[:]
    kn_round = 0
    while len(current) > 1:
        kn_round += 1
        next_round = []
        round_matches = []
        for i in range(0, len(current), 2):
            a, b = current[i], current[i + 1]
            round_matches.append((a, b))
            next_round.append(f"KR{kn_round}-M{i//2 + 1}-W")  # winner placeholder

        # Tournament matches use (courts - 1) during knockout if possible, so that
        # at least 1 court can be reserved for eliminated-teams' friendly matches.
        # If courts == 1, no reservation possible.
        tournament_courts = max(1, num_courts - 1) if num_courts > 1 else 1

        for bs in range(0, len(round_matches), tournament_courts):
            batch = round_matches[bs : bs + tournament_courts]
            if slot_idx >= max_slots:
                return {"fits": False, "reason": "淘汰赛超时"}
            matches = [
                {"court": i + 1, "team_a": a, "team_b": b, "kind": "ranked"}
                for i, (a, b) in enumerate(batch)
            ]
            # Reserve last court for friendlies if we have extra capacity
            if num_courts > len(batch):
                matches.append({
                    "court": num_courts,
                    "team_a": "自由",
                    "team_b": "自由",
                    "kind": "friendly",
                })
            schedule.append({
                "phase": "knockout",
                "round": f"KR{kn_round}",
                "slot": slot_idx + 1,
                "matches": matches,
            })
            slot_idx += 1

        current = next_round

    return {
        "format": "groups-knockout",
        "group_sizes": group_sizes,
        "knockout_size": kn_size,
        "schedule": schedule,
        "fits": True,
        "slots_used": slot_idx,
    }


def recommend_format(teams, num_courts, match_duration, total_time):
    """Pick the best format automatically."""
    T = len(teams)
    if T < 2:
        return {"error": "至少需要2支队伍", "fits": False}

    # Prefer groups + knockout if T >= 4
    if T >= 4:
        gk = plan_groups_knockout(teams, num_courts, match_duration, total_time)
        if gk.get("fits"):
            return gk

    # Fall back to round-robin
    return plan_round_robin(teams, num_courts, match_duration, total_time)


# ────────────────────────────────────────────────────────────────────────────
# Pretty printer
# ────────────────────────────────────────────────────────────────────────────

def format_plan(plan: dict, teams: list[dict], match_duration: int) -> str:
    if plan.get("error"):
        return f"Error: {plan['error']}"
    if not plan.get("fits", True):
        return f"Error: {plan.get('reason', 'unknown')}"

    lines = []
    lines.append("=" * 54)
    lines.append(f"FORMAT: {plan['format']}")
    if plan.get("group_sizes"):
        lines.append(f"Groups: {plan['group_sizes']}  Knockout: {plan['knockout_size']}")
    lines.append(f"Slots used: {plan['slots_used']} (duration per slot: {match_duration}m)")
    lines.append("=" * 54)

    def team_name(x):
        if isinstance(x, int):
            return teams[x]["name"]
        return x  # placeholder string

    for slot in plan["schedule"]:
        header = f"[Slot {slot['slot']}] {slot['phase']} round {slot['round']}"
        lines.append(header)
        for m in slot["matches"]:
            tag = "" if m["kind"] == "ranked" else " (friendly)"
            lines.append(
                f"  Court {m['court']}: {team_name(m['team_a'])} vs {team_name(m['team_b'])}{tag}"
            )
    return "\n".join(lines)


# ────────────────────────────────────────────────────────────────────────────
# Demo
# ────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    # Seed a fake database
    db = {}
    for i, name in enumerate([
        "Alice", "Bob", "Charlie", "Diana", "Eve", "Frank",
        "Grace", "Henry", "Ivy", "Jack", "Kate", "Leo",
    ]):
        db[name] = new_player(name)
        db[name]["points"] = (12 - i) * 3  # descending points for demo

    attendees = list(db.keys())

    # Form teams
    res = form_balanced_teams(attendees, db, team_size=2, seed=42)
    print("=" * 54)
    print("BALANCED TEAMS")
    print("=" * 54)
    for t in res["teams"]:
        names = ", ".join(t["players"])
        print(f"  {t['name']}: {names}")
    print()

    # Recommend format
    plan = recommend_format(
        teams=res["teams"],
        num_courts=2,
        match_duration=15,
        total_time=180,
    )
    print(format_plan(plan, res["teams"], match_duration=15))
